instrumentation-api/src/test/java/io/opentelemetry/instrumentation/api/instrumenter/messaging/MessagingAttributesExtractorTest.java

```diff
@@ -0,0 +1,178 @@
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+import org.assertj.core.data.MapEntry;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+class MessagingAttributesExtractorTest {
+  MessagingAttributesExtractor<Map<String, String>, String> underTest =
+      new MessagingAttributesExtractor<Map<String, String>, String>() {
+        @Override
+        protected String system(Map<String, String> request) {
+          return request.get("system");
+        }
+
+        @Override
+        protected String destinationKind(Map<String, String> request) {
+          return request.get("destinationKind");
+        }
+
+        @Override
+        protected String destination(Map<String, String> request) {
+          return request.get("destination");
+        }
+
+        @Override
+        protected boolean temporaryDestination(Map<String, String> request) {
+          return request.containsKey("temporaryDestination");
+        }
+
+        @Override
+        protected String protocol(Map<String, String> request) {
+          return request.get("protocol");
+        }
+
+        @Override
+        protected String protocolVersion(Map<String, String> request) {
+          return request.get("protocolVersion");
+        }
+
+        @Override
+        protected String url(Map<String, String> request) {
+          return request.get("url");
+        }
+
+        @Override
+        protected String conversationId(Map<String, String> request) {
+          return request.get("conversationId");
+        }
+
+        @Override
+        protected Long messagePayloadSize(Map<String, String> request) {
+          String payloadSize = request.get("payloadSize");
+          return payloadSize == null ? null : Long.valueOf(payloadSize);
+        }
+
+        @Override
+        protected Long messagePayloadCompressedSize(Map<String, String> request) {
+          String payloadSize = request.get("payloadCompressedSize");
+          return payloadSize == null ? null : Long.valueOf(payloadSize);
+        }
+
+        @Override
+        protected MessageOperation operation(Map<String, String> request) {
+          String operation = request.get("operation");
+          return operation == null ? null : MessageOperation.valueOf(operation);
+        }
+
+        @Override
+        protected String messageId(Map<String, String> request, String response) {
+          return response;
+        }
+      };
+
+  @ParameterizedTest
+  @MethodSource("destinations")
+  void shouldExtractAllAvailableAttributes(
```

Comment:
I am a little bit sceptical about the value of this test. It _may_ be useful to test extracting operation name and behaviour in case of temporary destination. But otherwise, maybe it can be replaced with tests of actual implementations of `MessagingAttributesExtractor`?

Reply:
I think that the real value is testing the full semantic conventions themselves, not just one implementation of them.

Maybe if we replaced the inheritance with composition too those tests would make more sense: something like `MessagingAttributesExtractor.create(new JmsAttributesGetter())`

Reply:
I refactored the attributes extractor and made it not abstract: https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/2803/commits/73f848ae595f582b48d9bdfe7dd1240ac4cec019
Please tell me WDYT about it - the separation of concerns is nice, but on the other hand it makes setting up an `Instrumenter` a bit more bloated.

Reply:
Hmm - I guess I lean towards the abstract version if it keeps using the `Instrumenter` simpler. Not sure there was a huge problem with the test

Reply:
I'll revert it for now, we can come back to this discussion in a separate issue/PR. Personally I don't really have a strong opinion about either solution, both have some good and bad sides.